### ISSUE TITLE
[Bug] word level timestamps with zero duration

### ISSUE THREAD

**Issue by jozefchutka** [xenova/transformers.js#223]
Hi, I just played with the recently released https://github.com/xenova/transformers.js/issues/68#issuecomment-1627851308 `return_timestamps: 'word'`. Mostly it produces solid transcription, however, in some instances returned words timestamps have 0 duration, meaning start=end

Here is an example:

```
27.4 -> 27.4  You
27.4 -> 27.54  have
27.54 -> 27.66  your
27.66 -> 28.02  robotics
...
39.56 -> 39.56  I
39.56 -> 39.56 'm
39.56 -> 39.68  having
39.68 -> 39.9  nightmares
...
42.88 -> 42.88  Oh,
42.88 -> 42.88  what
42.88 -> 42.88  in
42.88 -> 43  our
```

I am using following script:

```js
import { env, pipeline } from "https://cdn.jsdelivr.net/npm/@xenova/transformers@2.4.3/dist/transformers.min.js";

env.allowLocalModels = false;

const file = "tos.pcm";
const model = "Xenova/whisper-base.en";
const revision = "output_attentions";
const buffer = new Float32Array(await (await fetch(file)).arrayBuffer());
const pipe = await pipeline("automatic-speech-recognition", model, {revision});

const t0 = performance.now();
const result = await pipe(buffer, {
	chunk_length_s: 30,
	stride_length_s: 5,
	return_timestamps: "word"});

for(let {text, timestamp} of result.chunks)
	console.log(`${timestamp[0]} -> ${timestamp[1]} ${text}`);

console.log(performance.now() - t0);
```

I wonder if this is related to pipe configuration params or something else potentially fixable on transformers.js level or to be reported elsewhere?

**Comment by xenova** [comment]
Could you try the unquantized model? Also, since whisper wasn't actually designed for token-level timestamps, it's very possible that the approach which we implemented (dynamic time warping) might make a few mistakes here and there. If you try with the python library, you should get the same output. If you don't, then I'll look into it :)

**Comment by xenova** [comment]
Here's the original PR: https://github.com/huggingface/transformers/pull/23205#issuecomment-1561371887 (with a colab notebook on how to run it).

**Comment by jozefchutka** [comment]
I tried the notebook modified one step following https://gist.github.com/hollance/42e32852f24243b748ae6bc1f985b13a

```
pipe = pipeline(model="openai/whisper-base.en")

alignment_heads = [[3, 3], [4, 7], [5, 1], [5, 5], [5, 7]]
pipe.model.generation_config.alignment_heads = alignment_heads
```

Using the same source (.mp3 instead of .pcm) [tos.zip](https://github.com/xenova/transformers.js/files/12192242/tos.zip)

it produces:

```
  {'text': ' You', 'timestamp': (27.42, 27.42)},
  {'text': ' have', 'timestamp': (27.42, 27.54)},
  {'text': ' your', 'timestamp': (27.54, 27.66)},
  {'text': ' robotics', 'timestamp': (27.66, 28.0)},
...
{'text': " I'm", 'timestamp': (39.48, 39.54)},
  {'text': ' having', 'timestamp': (39.54, 39.64)},
  {'text': ' nightmares', 'timestamp': (39.64, 39.9)},
...
{'text': ' Oh', 'timestamp': (43.04, 43.04)},
  {'text': ' wait,', 'timestamp': (43.04, 43.04)},
  {'text': " we're", 'timestamp': (43.04, 43.04)},
  {'text': ' done!', 'timestamp': (43.04, 43.58)},
```

Interesting findings:

- timestamps are better in 2/3 cases, yet not perfect
- using same model and params (afaik) it produces better text like "I'm" comes as a single word unlike in .js where its two
- accuracy is also better

**Comment by xenova** [comment]
> .mp3 instead of .pcm

This might cause some minor differences, but not a big deal.

> timestamps are better in 2/3 cases, yet not perfect

👍 

> "I'm" comes as a single word unlike in .js where its two

Will look into that - it should be merged :)

**Comment by xenova** [comment]
Fixed the punctuation issue [here](https://github.com/xenova/transformers.js/pull/253/commits/d600ad61805c826e49ac10dd0effc895f4b9c5c0). This should actually fix many of the zero-duration timestamps (since they are merged into one now).

As for the others (and as you've seen when running the python version), those are just limitations of the algorithm. In future, it might be worth improving the python version, and when that is done, I could bring those changes here too.

**Comment by jozefchutka** [comment]
Hi @xenova , should "punctuation issue" be joining `I'm` together as a single? Using 2.5.2 , I am still seeing it as two.

**Comment by xenova** [comment]
Yes that will be fixed in 2.5.3 (which I still need to release)! 🤗

**Comment by xenova** [comment]
Released! 🥳 https://www.npmjs.com/package/@xenova/transformers/v/2.5.3
